File: 012/problem012.py
#!/usr/bin/env python
import math


def triangle_number(x):
    return sum(range(0, x + 1))


def square_root_as_int(x):
    return int(math.sqrt(x))


def factors(number):
    factors = set()
    for x in range(1, square_root_as_int(number) + 1):
        if number % x == 0 and number != x:
            factors.add(number / x)
            factors.add(x)
    return factors


x = 1
while len(factors(triangle_number(x))) < 500:
    x += 1

print(triangle_number(x))

File: src/012/problem012.rb
#!/usr/bin/env ruby
class Integer
  def triangle_number
    (1..self).reduce(:+)
  end

  def factors
    (1..Math.sqrt(self)).select { |i| (self % i).zero? }.each_with_object([]) do |number, list|
      list << number
      list << self / number unless self / number == number
    end
  end
end

x = 1
x += 1 while x.triangle_number.factors.count < 500
puts x.triangle_number
